File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseAccessibilityReviewLogic.kt
package com.lextures.android.core.lms

import kotlin.math.roundToInt

/** Course accessibility / alt-text review helpers (M13.8). */
object CourseAccessibilityReviewLogic {
    const val DECORATIVE_TITLE_MARKER = "lex-decorative"
    const val PAGE_SIZE = 20

    private val markdownImagePattern =
        Regex("""!\[([^\]]*)\]\(([^)\s]+)(?:\s+"([^"]*)")?\)""")

    data class MarkdownImageRef(
        val globalIndex: Int,
        val alt: String,
        val src: String,
        val title: String?,
        val decorative: Boolean,
        val hasValidAlt: Boolean,
        val line: Int,
    )

    data class ImageAltDraft(
        val alt: String,
        val decorative: Boolean,
    )

    data class AltTextUpdate(
        val imageIndex: Int,
        val alt: String,
        val decorative: Boolean,
    )

    fun cacheKey(courseCode: String): String = "course:$courseCode:accessibility"

    fun saveMarkdownIdempotencyKey(courseCode: String, itemId: String, kind: String): String =
        "course-accessibility:$courseCode:$kind:$itemId:markdown"

    fun markdownPatchPath(courseCode: String, itemId: String, kind: String): String? = when (kind) {
        "content_page" -> "/api/v1/courses/$courseCode/content-pages/$itemId"
        "assignment" -> "/api/v1/courses/$courseCode/assignments/$itemId"
        else -> null
    }

    fun supportsInlineEdit(kind: String): Boolean = kind == "content_page" || kind == "assignment"

    fun coveragePercent(withAlt: Int, total: Int): Int {
        if (total <= 0) return 100
        return ((withAlt.toDouble() / total.toDouble()) * 100).roundToInt()
    }

    fun paginatedUncoveredItems(items: List<UncoveredAccessibilityItem>, page: Int): List<UncoveredAccessibilityItem> {
        val end = minOf(items.size, maxOf(0, page + 1) * PAGE_SIZE)
        return items.take(end)
    }

    fun hasMorePages(items: List<UncoveredAccessibilityItem>, page: Int): Boolean =
        items.size > (page + 1) * PAGE_SIZE

    fun scanMarkdownImages(markdown: String): List<MarkdownImageRef> {
        val refs = mutableListOf<MarkdownImageRef>()
        var globalIndex = 0
        markdown.split("\n").forEachIndexed { lineOffset, line ->
            var searchStart = 0
            while (searchStart < line.length) {
                val match = markdownImagePattern.find(line, searchStart) ?: break
                val alt = match.groupValues[1]
                val src = match.groupValues[2]
                val title = match.groups[3]?.value
                val decorative = title == DECORATIVE_TITLE_MARKER
                val hasValidAlt = decorative || alt.trim().isNotEmpty()
                refs += MarkdownImageRef(
                    globalIndex = globalIndex,
                    alt = alt,
                    src = src,
                    title = title,
                    decorative = decorative,
                    hasValidAlt = hasValidAlt,
                    line = lineOffset + 1,
                )
                globalIndex += 1
                searchStart = match.range.last + 1
            }
        }
        return refs
    }

    fun missingImages(markdown: String): List<MarkdownImageRef> =
        scanMarkdownImages(markdown).filter { !it.hasValidAlt }

    fun applyAltTextUpdate(
        markdown: String,
        imageIndex: Int,
        alt: String,
        decorative: Boolean,
    ): String? = applyAltTextUpdates(markdown, listOf(AltTextUpdate(imageIndex, alt, decorative)))

    fun applyAltTextUpdates(markdown: String, updates: List<AltTextUpdate>): String? {
        if (updates.isEmpty()) return markdown
        val matches = markdownImagePattern.findAll(markdown).toList()
        var result = markdown
        for (update in updates.sortedByDescending { it.imageIndex }) {
            if (update.imageIndex !in matches.indices) return null
            val match = matches[update.imageIndex]
            val src = match.groupValues[2]
            val replacement = if (update.decorative) {
                "![]($src \"$DECORATIVE_TITLE_MARKER\")"
            } else {
                val trimmedAlt = update.alt.trim().replace("]", "\\]")
                "![$trimmedAlt]($src)"
            }
            result = result.replaceRange(match.range, replacement)
        }
        return result
    }

    fun kindLabelRes(kind: String): Int = when (kind) {
        "assignment" -> com.lextures.android.R.string.mobile_courseSettings_accessibility_kind_assignment
        "content_page" -> com.lextures.android.R.string.mobile_courseSettings_accessibility_kind_contentPage
        else -> com.lextures.android.R.string.mobile_courseSettings_accessibility_kind_other
    }

    fun drafts(images: List<MarkdownImageRef>): Map<Int, ImageAltDraft> =
        images.associate { it.globalIndex to ImageAltDraft(it.alt, it.decorative) }

    fun pendingUpdates(
        images: List<MarkdownImageRef>,
        drafts: Map<Int, ImageAltDraft>,
    ): List<AltTextUpdate> = images.mapNotNull { image ->
        val draft = drafts[image.globalIndex] ?: return@mapNotNull null
        val trimmedAlt = draft.alt.trim()
        val resolved = draft.decorative || trimmedAlt.isNotEmpty()
        if (!resolved) return@mapNotNull null
        if (draft.decorative == image.decorative && trimmedAlt == image.alt.trim()) return@mapNotNull null
        AltTextUpdate(image.globalIndex, trimmedAlt, draft.decorative)
    }
}

File: clients/ios/Lextures/Core/LMS/CourseAccessibilityReviewLogic.swift
import Foundation

/// Course accessibility / alt-text review helpers (M13.8).
enum CourseAccessibilityReviewLogic {
    static let decorativeTitleMarker = "lex-decorative"
    static let pageSize = 20

    struct MarkdownImageRef: Identifiable, Equatable, Hashable {
        var globalIndex: Int
        var alt: String
        var src: String
        var title: String?
        var decorative: Bool
        var hasValidAlt: Bool
        var line: Int

        var id: Int { globalIndex }
    }

    struct ImageAltDraft: Equatable, Hashable {
        var alt: String
        var decorative: Bool
    }

    static func cacheKey(courseCode: String) -> String {
        "course:\(courseCode):accessibility"
    }

    static func saveMarkdownIdempotencyKey(courseCode: String, itemId: String, kind: String) -> String {
        "course-accessibility:\(courseCode):\(kind):\(itemId):markdown"
    }

    static func markdownPatchPath(courseCode: String, itemId: String, kind: String) -> String? {
        switch kind {
        case "content_page":
            return "/api/v1/courses/\(courseCode)/content-pages/\(itemId)"
        case "assignment":
            return "/api/v1/courses/\(courseCode)/assignments/\(itemId)"
        default:
            return nil
        }
    }

    static func supportsInlineEdit(kind: String) -> Bool {
        kind == "content_page" || kind == "assignment"
    }

    static func coveragePercent(withAlt: Int, total: Int) -> Int {
        if total <= 0 { return 100 }
        return Int((Double(withAlt) / Double(total) * 100).rounded())
    }

    static func formatCoverageLabel(withAlt: Int, total: Int) -> String {
        let pct = coveragePercent(withAlt: withAlt, total: total)
        return L.format("mobile.courseSettings.accessibility.coverageValue", pct, withAlt, total)
    }

    static func paginatedUncoveredItems(_ items: [UncoveredAccessibilityItem], page: Int) -> [UncoveredAccessibilityItem] {
        let end = min(items.count, max(0, page + 1) * pageSize)
        return Array(items.prefix(end))
    }

    static func hasMorePages(items: [UncoveredAccessibilityItem], page: Int) -> Bool {
        items.count > (page + 1) * pageSize
    }

    static func scanMarkdownImages(_ markdown: String) -> [MarkdownImageRef] {
        let pattern = /!\[([^\]]*)\]\(([^)\s]+)(?:\s+"([^"]*)")?\)/
        var refs: [MarkdownImageRef] = []
        var globalIndex = 0
        let lines = markdown.split(separator: "\n", omittingEmptySubsequences: false)
        for (lineOffset, lineSub) in lines.enumerated() {
            let line = String(lineSub)
            var searchStart = line.startIndex
            while searchStart < line.endIndex {
                guard let match = line[searchStart...].firstMatch(of: pattern) else { break }
                let alt = String(match.output.1)
                let src = String(match.output.2)
                let title = match.output.3.map(String.init)
                let decorative = title == decorativeTitleMarker
                let hasValidAlt = decorative || !alt.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
                refs.append(MarkdownImageRef(
                    globalIndex: globalIndex,
                    alt: alt,
                    src: src,
                    title: title,
                    decorative: decorative,
                    hasValidAlt: hasValidAlt,
                    line: lineOffset + 1
                ))
                globalIndex += 1
                searchStart = match.range.upperBound
            }
        }
        return refs
    }

    static func missingImages(_ markdown: String) -> [MarkdownImageRef] {
        scanMarkdownImages(markdown).filter { !$0.hasValidAlt }
    }

    static func applyAltTextUpdate(
        in markdown: String,
        imageIndex: Int,
        alt: String,
        decorative: Bool
    ) -> String? {
        applyAltTextUpdates(in: markdown, updates: [(imageIndex: imageIndex, alt: alt, decorative: decorative)])
    }

    static func applyAltTextUpdates(
        in markdown: String,
        updates: [(imageIndex: Int, alt: String, decorative: Bool)]
    ) -> String? {
        guard !updates.isEmpty else { return markdown }
        let pattern = try? NSRegularExpression(pattern: #"!\[([^\]]*)\]\(([^)\s]+)(?:\s+"([^"]*)")?\)"#)
        guard let pattern else { return nil }
        let ns = markdown as NSString
        let matches = pattern.matches(in: markdown, range: NSRange(location: 0, length: ns.length))
        var result = markdown
        for update in updates.sorted(by: { $0.imageIndex > $1.imageIndex }) {
            guard update.imageIndex >= 0, update.imageIndex < matches.count else { return nil }
            let match = matches[update.imageIndex]
            let src = ns.substring(with: match.range(at: 2))
            let replacement: String
            if update.decorative {
                replacement = "![](\(src) \"\(decorativeTitleMarker)\")"
            } else {
                let trimmedAlt = update.alt.trimmingCharacters(in: .whitespacesAndNewlines)
                let escapedAlt = trimmedAlt.replacingOccurrences(of: "]", with: "\\]")
                replacement = "![\(escapedAlt)](\(src))"
            }
            guard let range = Range(match.range, in: result) else { return nil }
            result.replaceSubrange(range, with: replacement)
        }
        return result
    }

    static func kindLabelKey(for kind: String) -> String {
        switch kind {
        case "assignment": return "mobile.courseSettings.accessibility.kind.assignment"
        case "content_page": return "mobile.courseSettings.accessibility.kind.contentPage"
        default: return "mobile.courseSettings.accessibility.kind.other"
        }
    }

    static func itemMissingLabel(missing: Int, total: Int) -> String {
        L.format("mobile.courseSettings.accessibility.itemMissing", missing, total)
    }

    static func drafts(from images: [MarkdownImageRef]) -> [Int: ImageAltDraft] {
        Dictionary(uniqueKeysWithValues: images.map { image in
            (image.globalIndex, ImageAltDraft(alt: image.alt, decorative: image.decorative))
        })
    }

    static func pendingUpdates(
        images: [MarkdownImageRef],
        drafts: [Int: ImageAltDraft]
    ) -> [(imageIndex: Int, alt: String, decorative: Bool)] {
        images.compactMap { image in
            guard let draft = drafts[image.globalIndex] else { return nil }
            let trimmedAlt = draft.alt.trimmingCharacters(in: .whitespacesAndNewlines)
            let resolved = draft.decorative || !trimmedAlt.isEmpty
            guard resolved else { return nil }
            if draft.decorative == image.decorative && trimmedAlt == image.alt.trimmingCharacters(in: .whitespacesAndNewlines) {
                return nil
            }
            return (imageIndex: image.globalIndex, alt: trimmedAlt, decorative: draft.decorative)
        }
    }
}
